src/main/java/com/google/cloud/tools/appengine/v2/internal/process/CliProcess.java

```diff
@@ -0,0 +1,223 @@
+import com.google.cloud.tools.appengine.v2.internal.cloudsdk.StringResultConverter;
+import com.google.cloud.tools.appengine.v2.OutputHandler;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFutureTask;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Scanner;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * The cli process mechanism starts a new process and starts 3 threads on an executor
+ * 1. The processMain thread waits on the process to end and stores the exit code
+ * 2. The StdErr thread processes the err stream and writes output to the outputListener
+ * 3. The StdOut thread processes the out stream and stores output to return as result
+ *
+ * The future implementation considers the process main (1) thread to be the primary thread
+ * to watch. It forwards all future interface calls to the future returned when starting that
+ * thread (1).
+ *
+ * @param <T> the process return type
+ */
+public class CliProcess<T> implements Future<T> {
+
+  private final Map<String, String> environment;
+  private final ExecutorService executor;
+  private final OutputHandler outputHandler;
+  private final StringResultConverter<T> stringResultConverter;
+  private final String[] command;
+  private final AppEngineRequestCallback<T> callback;
+  private Process process;
+  private ListenableFutureTask<CliProcessResult<T>> processMain;
+  private ListenableFutureTask<String> processStdOut;
+  private Runnable processStdErr;
+
+  private CliProcess(OutputHandler outputHandler, Map<String, String> environment,
+      String[] command, StringResultConverter<T> stringResultConverter,
+      AppEngineRequestCallback<T> callback) {
+
+    this.executor = MoreExecutors.listeningDecorator(MoreExecutors.getExitingExecutorService(
+        (ThreadPoolExecutor) Executors.newFixedThreadPool(3), 2, TimeUnit.SECONDS));
+    this.outputHandler = outputHandler;
+    this.environment = environment;
+    this.command = command;
+    this.stringResultConverter = stringResultConverter;
+    this.callback = callback;
+  }
+
+  // Main entry point, starts process and adds three threads to the executor
+  private Future<T> start() throws IOException {
+    //ProcessBuilder pb = new ProcessBuilder("/tmp/test.sh");
+    ProcessBuilder pb = new ProcessBuilder(command);
+    pb.environment().putAll(environment);
+    process = pb.start();
+
+
+    processStdOut = ListenableFutureTask.create(new Callable<String>() {
```

Comment:
I think what we might do is a raw Process option that just returns access to a raw process instead of a managed request. So the user can do one or the other.